clients/adapter/src/main/java/org/eclipse/hono/adapter/client/command/CommandRouterCommandConsumerFactory.java

```diff
@@ -11,100 +11,111 @@
-    private final ProtonBasedAdapterInstanceCommandHandler adapterInstanceCommandHandler;
-    private final AtomicBoolean recreatingConsumer = new AtomicBoolean(false);
-    private final AtomicBoolean tryAgainRecreatingConsumer = new AtomicBoolean(false);
-
     private final CommandRouterClient commandRouterClient;
-    private ProtonReceiver adapterSpecificConsumer;
+    private final List<Lifecycle> internalCommandConsumers = new ArrayList<>();
 
     /**
-     * Creates a new factory for an existing connection.
+     * Creates a new factory.
      *
-     * @param connection The connection to the AMQP network.
-     * @param samplerFactory The sampler factory to use.
      * @param commandRouterClient The client to use for accessing the command router service.
+     * @param adapterName The name of the protocol adapter.
      * @throws NullPointerException if any of the parameters is {@code null}.
      */
-    public ProtonBasedCommandRouterCommandConsumerFactoryImpl(
-            final HonoConnection connection,
-            final SendMessageSampler.Factory samplerFactory,
-            final CommandRouterClient commandRouterClient) {
-        super(connection, samplerFactory);
+    public CommandRouterCommandConsumerFactory(final CommandRouterClient commandRouterClient, final String adapterName) {
         this.commandRouterClient = Objects.requireNonNull(commandRouterClient);
-
-        adapterInstanceId = getAdapterInstanceId(connection.getConfig().getName());
-        adapterInstanceCommandHandler = new ProtonBasedAdapterInstanceCommandHandler(connection.getTracer(), adapterInstanceId);
+        this.adapterInstanceId = getNewAdapterInstanceId(Objects.requireNonNull(adapterName));
     }
 
-    private static String getAdapterInstanceId(final String adapterName) {
+    private static String getNewAdapterInstanceId(final String adapterName) {
         // replace special characters so that the id can be used in a Kafka topic name
         final String prefix = Strings.isNullOrEmpty(adapterName) ? ""
                 : adapterName.replaceAll("[^a-zA-Z0-9._-]", "") + "-";
         return prefix + UUID.randomUUID();
     }
 
+    /**
+     * Registers the command consumer receiving commands on the internal command endpoint of the protocol adapter.
+     * That is the endpoint that the Command Router forwards received commands to.
+     * <p>
+     * Note that this method needs to be called before invoking {@link #start()}. The {@link #start()} and
+     * {@link #stop()} methods of this factory will invoke the corresponding methods on the internal command consumer.
+     *
+     * @param internalCommandConsumerSupplier Function that returns the consumer. Parameters are the adapter instance
+     *            id that identifies the internal command endpoint and the command handlers to choose from when handling
+     *            a received command.
+     */
+    public void registerInternalCommandConsumer(
```

Comment:
is this method intended to be invoked once only or several times?

Reply:
It's intended to be invoked once only if the Hono deployment uses an AMQP messaging solution *or* a Kafka cluster.
If both kinds of messaging solutions are configured for the adapter, it's intended to be invoked twice, once for each Kafka or Proton specific internal consumer implementation.

Reply:
got it, thanks for explaining